### ISSUE TITLE
Value is always True/False warning

### ISSUE THREAD

**Issue by kendrahavens** [dotnet/roslyn#33164]
**Version Used**:  Visual Studio 2019 Preview 2.1

Copying customer request

**Steps to Reproduce**:
We get a grayed out "unreachable code" warning if code is unreachable when a condition is always false, but not when a condition is always true.
```csharp
        private const int V = 100;

        public void AlwaysFalse()
        {
            if(V != 100)
            {
                Console.WriteLine("is 100" + connnection);
            }

        }
```

So if I have the below condition that will always be true. The customer is requesting a warning:
```csharp
        private const int V = 100;

        public void AlwaysTrue()
        {
            if(V == 100)
            {
                Console.WriteLine("is 100" + connnection);
            }

        }

```

**Expected Behavior**:
Warning on conditions that are always true.

**Actual Behavior**:
No warning

**Comment by mavasani** [comment]
We have a dataflow analysis based implementation for this in the Roslyn-analyzers repo: https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/AvoidDeadConditionalCode.cs

It detects conditions which are always true/false based on flow analysis of constant and non-constant values, and also uses copy analysis to track variables which have same values.

See unit tests https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs and https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_NullAnalysis.cs for detailed examples.

**Comment by mavasani** [comment]
For example, see the [below test](https://github.com/dotnet/roslyn-analyzers/blob/features/dataflow/src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/AvoidDeadConditionalCode_ValueContentAnalysis.cs#L113):
```csharp
        [Trait(Traits.DataflowAnalysis, Traits.Dataflow.ValueContentAnalysis)]
        [Fact]
        public void ValueCompareWithSubtract_Diagnostic()
        {
            VerifyCSharp(@"
class Test
{
    void M1(int param, int param2, int param3)
    {
        param2 = 2;
        if (param3 == 3 && (param3 - param2 == param))
        {
            if (param == 1)
            {
            }
        }
    }
}
",
            // Test0.cs(9,17): warning CA1508: 'param == 1' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
            GetCSharpResultAt(9, 17, "param == 1", "true"));
```

**Comment by jasonmalinowski** [comment]
@mavasani @sharwell This is marked as "needs design review", but it's unclear to me what the questions are.

**Comment by mavasani** [comment]
I think we already discussed this at the design meeting, and given we have an implementation already in the analyzers repo (part of FxCop analyzers package), we had decided to resolve this one as external.

**Comment by jasonmalinowski** [comment]
@mavasani So you think this work item just wasn't updated?

**Comment by mavasani** [comment]
Yes, I believe so. I am not sure if Sam had anything else to add from his notes.

**Comment by sharwell** [comment]
We reviewed this 6 May 2019. Since this feature is already implemented in dotnet/roslyn-analyzers, we are marking it **External**.
